src/config.rs

```diff
@@ -190,28 +192,126 @@ impl LayeredConfigs {
     }
 }
 
-pub fn config_path() -> Result<Option<PathBuf>, ConfigError> {
-    if let Ok(config_path) = env::var("JJ_CONFIG") {
-        // TODO: We should probably support colon-separated (std::env::split_paths)
-        // paths here
-        Ok(Some(PathBuf::from(config_path)))
-    } else {
+enum ConfigPath {
+    /// Path to an existing config file. Note that this is not necessarily
+    /// checked (eg. if specified via the JJ_CONFIG env var).
+    Existing(PathBuf),
+    /// Paths to potential locations of the new config file, in the order of
+    /// preference.
+    New(PathBuf),
+    Unavailable,
+}
+
+impl ConfigPath {
+    fn new(path: Option<PathBuf>) -> Self {
+        match path {
+            Some(path) if path.exists() => ConfigPath::Existing(path),
+            Some(path) => ConfigPath::New(path),
+            None => ConfigPath::Unavailable,
+        }
+    }
+}
+
+fn create_config_file(path: &Path) -> std::io::Result<std::fs::File> {
+    if let Some(parent) = path.parent() {
+        std::fs::create_dir_all(parent)?;
+    }
+    // TODO: Use File::create_new once stabilized.
+    std::fs::OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create_new(true)
+        .open(path)
+}
+
+// The struct exists so that we can mock certain global values in unit tests.
+#[derive(Clone, Default, Debug)]
+struct ConfigEnv {
+    config_dir: Option<PathBuf>,
+    home_dir: Option<PathBuf>,
+    jj_config: Option<String>,
+}
+
+impl ConfigEnv {
+    fn new() -> Self {
+        ConfigEnv {
+            config_dir: dirs::config_dir(),
+            home_dir: dirs::home_dir(),
+            jj_config: env::var("JJ_CONFIG").ok(),
+        }
+    }
+
+    fn config_path(self, allow_create: bool) -> Result<Option<PathBuf>, ConfigError> {
+        if let Some(path) = self.jj_config {
+            // TODO: We should probably support colon-separated (std::env::split_paths)
+            // paths here
+            let path = PathBuf::from(path);
+            if allow_create {
+                if let Err(e) = create_config_file(&path) {
+                    if e.kind() != std::io::ErrorKind::AlreadyExists {
+                        return Err(e.into());
+                    }
+                }
+            }
```

Comment:
Perhaps, `config_path()` can be a pure function that returns `Result<ConfigPath, _>` enum, and any side effect like creating empty config file can be moved to `new_config_path()`.

Reply:
Was thinking about it before, but the impl got a bit messy - though potentially because I wanted to fallback to using one directory if another failed to create. Will circle back here once we get clarity on the directory creation.